.github/workflows/stackhpc-container-image-build.yml

```diff
@@ -200,12 +208,69 @@ jobs:
       - name: Fail if no images have been built
         run: if [ $(wc -l < ${{ matrix.distro }}-container-images) -le 1 ]; then exit 1; fi
 
-      - name: Upload container images artifact
+      - name: Scan built container images
+        run: |
+          src/kayobe-config/tools/scan-images.sh ${{ matrix.distro }} ${{ needs.generate-tag.outputs.kolla_tag }}
+
+      - name: Fail if no images have passed scanning
+        run: if [ $(wc -l < image-scan-output/clean-images.txt) -le 0 ]; then exit 1; fi
+        if: ${{ !inputs.push-dirty }}
+
+      - name: Copy clean images to push-attempt-images list
+        run: cp image-scan-output/clean-images.txt image-scan-output/push-attempt-images.txt
+
+      - name: Append dirty images to push list
+        run: |
+          cat image-scan-output/dirty-images.txt >> image-scan-output/push-attempt-images.txt
+        if: inputs.push-dirty
+
+      - name: Push images
+        run: |
+          touch image-scan-output/push-failed-images.txt
+          source venvs/kayobe/bin/activate &&
+          source src/kayobe-config/kayobe-env --environment ci-builder &&
+          kayobe playbook run ${KAYOBE_CONFIG_PATH}/ansible/docker-registry-login.yml &&
+
+          while read -r image; do
+            # Retries!
+            for i in {1..5}; do 
+              if docker push $image; then
+                echo "Pushed $image"
+                break
+              elif $i == 5; then
+                echo "Failed to push $image"
+                echo $image >> image-scan-output/push-failed-images.txt
+              else
+                echo "Failed on retry $i"
+                sleep 5
+              fi;
+            done
+          done < image-scan-output/push-attempt-images.txt &&
+          mv image-scan-output image-build-logs
+        shell: bash
+        env:
+          KAYOBE_VAULT_PASSWORD: ${{ secrets.KAYOBE_VAULT_PASSWORD }}
+        if: inputs.push
+
+      - name: Upload output artifact
         uses: actions/upload-artifact@v4
         with:
-          name: ${{ matrix.distro }} container images
-          path: ${{ matrix.distro }}-container-images
+          name: ${{ matrix.distro }}-logs
+          path: image-build-logs
           retention-days: 7
+        if: !cancelled()
+
+      - name: Fail when images failed to build
+        run: exit 1
+        if: steps.build_overcloud_images.outcome == 'failure' || steps.build_seed_images.outcome == 'failure'
+
+      - name: Fail when images failed to push
+        run: if [ $(wc -l < image-build-logs/push-failed-images.txt) -gt 0 ]; then cat image-build-logs/push-failed-images.txt && exit 1; fi
+        if: !cancelled()
+
+      - name: Fail when images failed scanning
+        run: if [ $(wc -l < image-build-logs/dirty-images.txt) -gt 0 ]; then cat image-build-logs/dirty-images.txt && exit 1; fi
+        if: ${{ !inputs.push-dirty && !cancelled() }}
 
   sync-container-repositories:
```

Comment:
There is an issue that if the previous job fails this one won't run, so images won't be synced. That's not a problem for now (RIP SMS) but may become one again.

Reply:
One option could be to put these steps in a separate job that the sync job doesn't depend on. It would need to download the artifact.

Reply:
Given that it's just one task, could we integrate it into the main action, which would make it easier to control when it gets triggered

Reply:
That's another option.

Reply:
This one's still open. Otherwise looks good.

Reply:
Done